### PR TITLE
Support datetime functions: MONTH, DAYOFMONTH, DATE, MONTHNAME, TIMESTAMP, MAKETIME, NOW, CURDATE

### PR DESCRIPTION
**Pull request by chloe-zh**
*Issue #, if available:*

* [**Issue #235 Unable to recognize certain function names**](https://github.com/opendistro-for-elasticsearch/sql/issues/235)

*Description of changes:*

* Implemented datetime functions: MONTH, DAYOFMONTH, DATE, MONTHNAME, TIMESTAMP, MAKETIME, NOW, CURDATE.
* Added UT and IT.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java [hunk 1]
```diff
@@ -67,7 +67,8 @@
 
     private static final Set<String> dateFunctions = Sets.newHashSet(
             "date_format", "year", "month_of_year", "week_of_year", "day_of_year", "day_of_month",
-            "day_of_week", "hour_of_day", "minute_of_day", "minute_of_hour", "second_of_minute"
+            "day_of_week", "hour_of_day", "minute_of_day", "minute_of_hour", "second_of_minute", "month", "dayofmonth",
```
Comment (davidcui-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java]: the naming scheme seems inconsistent between having underscores between words (e.g dayofmonth vs second_of_minute)

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java [hunk 2]
```diff
@@ -183,6 +192,19 @@ public static boolean isFunctionTranslatedToScript(String function) {
...
+                functionStr = timestamp((SQLExpr) paramers.get(0).value);
+                break;
+            case "maketime":
+                functionStr = maketime((SQLExpr) paramers.get(0).value, (SQLExpr) paramers.get(1).value,
+                        (SQLExpr) paramers.get(2).value);
+                break;
+            case "now":
+                functionStr = now();
+                break;
+            case "curdate":
```
Comment (davidcui-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java]: I think this should be either currentdate or current_date; none of the other names seem to be truncated.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java [hunk 3]
```diff
@@ -91,7 +91,15 @@
...
-    YEAR(func(DATE).to(INTEGER));
+    YEAR(func(ESDataType.DATE).to(INTEGER)),
+    MONTH(func(ESDataType.DATE).to(INTEGER)),
+    MONTHNAME(func(ESDataType.DATE).to(STRING)),
+    DAYOFMONTH(func(ESDataType.DATE).to(INTEGER)),
+    DATE(func(ESDataType.DATE).to(ESDataType.DATE)),
+    TIMESTAMP(func(ESDataType.DATE).to(ESDataType.DATE)),
+    MAKETIME(func(INTEGER, INTEGER, INTEGER).to(ESDataType.DATE)),
+    NOW(),
+    CURDATE();
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java]: Should be `() -> DATE` ? None specification means it's unknown and we want to bypass any type check for now.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFunctionsTest.java [hunk 4]
```diff
@@ -127,4 +127,91 @@ public void secondOfMinute() {
                         "doc['creationDate'].value.secondOfMinute"));
     }
 
+    @Test
+    public void month() {
+        String query = "SELECT MONTH(creationDate) " +
+                "FROM dates";
```
Comment (dai-chen) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFunctionsTest.java]: could be one line?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java [hunk 5]
```diff
@@ -43,8 +43,8 @@
     COSH(func(T(NUMBER)).to(T)),
     COT(func(T(NUMBER)).to(T)),
     DATE_FORMAT(
-        func(DATE, STRING).to(STRING),
-        func(DATE, STRING, STRING).to(STRING)
+        func(ESDataType.DATE, STRING).to(STRING),
```
Comment (galkk) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java]: What is the reason behind this change?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java (hunk 1 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java]: The ones without underscore are supported by MySQL so we could put them into the scalar function base, while the underscored functions are ES_s. These function names are

Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java]: @davidcui-amzn Thanks! These functions are to support the BI integration.
The ones without underscore are supported by MySQL so we could put them into the scalar function base, while the underscored functions (if not supported by MySQL) are put into ES_Scalar functions.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java (hunk 5 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java]: Because we added a scalar function named `DATE` in the same class. This made some confusion to the "DATE" type.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java (hunk 2 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java]: This was referred from MySQL documentation, and also we need to support BI integration.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFunctionsTest.java (hunk 4 above)
Comment (chloe-zh) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFunctionsTest.java]: yes sure, will do.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java (hunk 3 above)
Comment (chloe-zh) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/function/ScalarFunction.java]: Done. Thanks!
